### DISCUSSION TITLE
How to check offline sync status?

### DISCUSSION THREAD

**Discussion by frank06** [firebase/flutterfire#6523] [Q&A]
When a user of our app loses connectivity and creates documents, these are stored in the local cache to be later synchronized by the Firestore client.

How can I obtain the status of what's left to be sync'd?

So far our attempt is setting a listener on all collections:

```dart
final _sub = StreamGroup.merge([
      for (final name in kTypes)
        db.collection(name).snapshots(includeMetadataChanges: true),
    ]).listen((qs) {
  // check for isFromCache / hasPendingWrites
});
```

However this ends up taking an unacceptably long time (~25 seconds on my fast laptop). Our collections are huge (100k+ documents) although they are consumed via queries elsewhere in the app. Not sure how many are being stored in cache.

Is there a more efficient way to check for documents that need to be synced?

**Comment by jamesdixon** [comment]
Would love to know this as well!

Not being able to provide our users feedback on what's synced and what hasn't is a really poor user experience. Would really appreciate a way to access the number of items left to sync both at a global and collection level.

cc: @samtstern @schmidt-sebastian

**Reply by schmidt-sebastian** [reply]
We currently do not have an API for this. We only expose `waitForPendingWrites`, but we do not offer a way to check the write status on a per document level other than through the Query API. You can also issue individual queries for each affected document, which should be much faster since we can directly read the document and do not have to scan your large collection.

**Reply by frank06** [reply]
Thanks. Can I listen for changes to multiple documents without merging streams? Or do I need to poll for updated metadata?

**Reply by schmidt-sebastian** [reply]
You have to listen to each document individually but you will not need to poll. Not that you can use a `source: cache` listener, which will not incur any additional billing.

**Reply by GabrielAraujo** [reply]
Hey @schmidt-sebastian is there a way to have realtime updates with snapshot filtering only from cache? I've found only the option to set the source only on `.get` and not on `.snapshots`

**Reply by schmidt-sebastian** [reply]
Not right now, but this has been an often requested feature. We have not yet come up with a timeline for this though.
